### PR TITLE
Adds typescript to config

### PR DESCRIPTION
**Pull request by jared-dykstra**
Improves the VSCode Prettier/Typescript/Eslint tooling experience.  I wasn't happy with the previous setup, particularly for `vanilla.template.ts`

Previously, VSCode's autofix would introduce changes in conflict with Prettier, such as adding spacing due to `rest-spread-spacing` rule

It's now more consistent.  Saving in VSCode, `prettier --write`, and `eslint --fix` are consistent.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  },
+  settings: {
+    react: {
+      version: 'detect',
+    },
+  },
+  rules: {
+    'prettier/prettier': 'error',
+  },
+};
```
Comment (jared-dykstra) [inline .eslintrc.js:25]: I don't know if we prefer `.js` or `.json`?  It's possible to put comments in the `.js` file without the `jsonc` workaround.

**Comment by jared-dykstra** [comment]
`@typescript-eslint/recommended` includes a rule to define return types for exports.  Inferred types are convenient, but rather than disabling that rule and straying from recommended defaults, I added extra definitions to the template.  Partners can make any ruleset changes they wish.

File: .eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+module.exports = {
+  plugins: ['prettier'],
+  parser: '@typescript-eslint/parser',
+  extends: [
+    'plugin:react/recommended',
+    'plugin:@typescript-eslint/recommended',
+    'prettier/@typescript-eslint',
```
Comment (lemonmade) [inline .eslintrc.js:7]: This doesn't need the `plugin:`?

File: .eslintrc.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+      jsx: true,
+    },
+  },
+  settings: {
+    react: {
+      version: 'detect',
+    },
+  },
+  rules: {
+    'prettier/prettier': 'error',
```
Comment (lemonmade) [inline .eslintrc.js:23]: This isn't in the prettier recommended rules?

File: scripts/generate/clean-up.ts [hunk 4]
```diff
@@ -2,7 +2,7 @@ import fs from 'fs';
 import path from 'path';
 import {exec} from 'child_process';
 
-export function cleanUp() {
+export function cleanUp(): void {
```
Comment (lemonmade) [inline scripts/generate/clean-up.ts:5]: 👎 to any rule that requires explicit, unnecessary type annotations. IMO all they do is make the a simple extension more confusing for beginners, when we already give them all the benefits of TypeScript without needing almost any explicit types.

Comment (jared-dykstra) [reply scripts/generate/clean-up.ts:5]: I do agree, that devs unaccustomed to TS have an extra burdon.  So, as per the [earlier comment](https://github.com/Shopify/argo-checkout-template/pull/27#issuecomment-654472306), I'll modify it and disable the rule, departing from vanilla / recommended setup.

File: .eslintrc.js (hunk 3 above)
Comment (jared-dykstra) [reply .eslintrc.js:23]: Good thought.  Will verify and remove if unnecessary

File: .eslintrc.js (hunk 2 above)
Comment (jared-dykstra) [reply .eslintrc.js:7]: yeah, probably not.  Will remove.  TY.

Comment (jared-dykstra) [reply .eslintrc.js:7]: To follow-up: `eslint-plugin-` prefix could be omitted, but the `plugin:` prefix [is needed](https://eslint.org/docs/user-guide/configuring#configuring-plugins)

File: .eslintrc.js (hunk 3 above)
Comment (jared-dykstra) [reply .eslintrc.js:23]: It is included by the recommended rules.  Can remove

File: scripts/generate/clean-up.ts (hunk 4 above)
Comment (jared-dykstra) [reply scripts/generate/clean-up.ts:5]: Disabled `@typescript-eslint/explicit-module-boundary-types'

**Pull request merged**